hadoop-ozone/client/src/test/java/org/apache/hadoop/ozone/client/MultiNodePipelineBlockAllocator.java

```diff
@@ -21,68 +21,103 @@
 import org.apache.hadoop.hdds.conf.ConfigurationSource;
 import org.apache.hadoop.hdds.conf.OzoneConfiguration;
 import org.apache.hadoop.hdds.conf.StorageUnit;
+import org.apache.hadoop.hdds.protocol.DatanodeDetails;
 import org.apache.hadoop.hdds.protocol.proto.HddsProtos;
+import org.apache.hadoop.hdds.scm.container.common.helpers.ExcludeList;
 import org.apache.hadoop.ozone.OzoneConfigKeys;
 import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos;
 
 import java.util.ArrayList;
 import java.util.List;
-import java.util.Random;
+import java.util.Set;
 
 /**
  * Allocates the block with required number of nodes in the pipeline.
  */
 public class MultiNodePipelineBlockAllocator implements MockBlockAllocator {
-  public static final Random RANDOM = new Random();
   private long blockId;
   private int requiredNodes;
   private final ConfigurationSource conf;
+  private List<HddsProtos.DatanodeDetailsProto> clusterDns = new ArrayList<>();
+  private int start = 0;
 
   public MultiNodePipelineBlockAllocator(OzoneConfiguration conf,
-      int requiredNodes) {
+      int requiredNodes, int clusterSize) {
```

Comment:
I think we should document a bit better how the block allocator works, as the code for it starts to get a bit more large to make it just easy to understand. It would be really useful to easily understand what is happening here, and how this code interacts with the failure injection logic inside MockXCieverClientFactory.

Reply:
I added the documentation for the mock allocatBlock API to explain how we choose nodes.